### ISSUE TITLE
panic: ConstrainEq on unbound variables

### ISSUE THREAD

**Issue by yihozhang** [egraphs-good/egglog#143]
```
(rule ((= x 1) (= y x) (= z y)) ())
(run 1)
```
panics, while
```
(rule ((= x 1) (= y x)) ())
(run 1)
```
does not.

Seems related to desugaring @oflatt?

This issue is blocking @clyben from fixing a follow-up issue of #93.

More details:

Currently, the program produces body
```
[AssignLit(x, 1), ConstrainEq(y, x), ConstrainEq(z, y)]
```
while the expected body is
```
[AssignLit(x, 1), AssignVar(y, x), AssignVar(z, y)]
```

(Note that we don't have AssignVar currently.)

`ConstraintEq`s are produced by `flatten_facts` and `flatten_equalities` on cases like `(= var1 var2)`.

We can also remove the checks since we are converting `NormFact::ConstrainEq` back to `Fact::Eq` anyway so it shouldn't matter, maybe?

**Comment by yihozhang** [comment]
Also just curious, why do we need to distinguish between `NormFact::Assign` and `NormFact::ConstrainEq`? Is it for proof generation? After all, the query compiler does not distinguish between them.

**Comment by oflatt** [comment]
Yeah I think the problem is that we disallow "unbound" variables in contrainEq. Variables need to be bound in an assignment
I'd rather keep this invariant- is there a way to work around it?

**Comment by yihozhang** [comment]
Fixed by #176.
